### PR TITLE
RESTWS-836: Add REST resource for Triggered State Conversions management

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
Added REST resource for Triggered State Conversions management

## Issue I worked on
see https://issues.openmrs.org/browse/RESTWS-836

PR with documentation: https://github.com/openmrs/openmrs-contrib-rest-api-docs/pull/158

## Checklist: I completed these to help reviewers :)
- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import org.openmrs.module.webservices.rest.web.resource.impl.DelegatingResourceDescription;
+import org.openmrs.module.webservices.rest.web.resource.impl.NeedsPaging;
+import org.openmrs.module.webservices.rest.web.response.IllegalRequestException;
+import org.openmrs.module.webservices.rest.web.response.ResourceDoesNotSupportOperationException;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+
+/**
+ * {@link Resource} for {@link ConceptStateConversion}, supporting standard CRUD operations
+ */
+@Resource(name = RestConstants.VERSION_1 + "/stateconversion", supportedClass = ConceptStateConversion.class,
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0.java:36]: `conceptstateconversion`?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0.java]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/493/commits/ae2e339bd5f16a18aa8b86f25838ad165a167fc2

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.openmrs.Concept;
+import org.openmrs.ConceptStateConversion;
+import org.openmrs.ProgramWorkflow;
+import org.openmrs.ProgramWorkflowState;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.webservices.rest.web.RestTestConstants1_8;
+import org.openmrs.module.webservices.rest.web.resource.impl.BaseDelegatingResourceTest;
+
+public class ConceptStateConversionResource2_0Test
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:20]: What is the use of this class?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:20]: Its responsibility is to make sure the returned resource representation has appropriate fields. I've added more assertions to test custom fields.

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+		ConceptStateConversion conceptStateConversion = new ConceptStateConversion();
+		conceptStateConversion.setConcept(concept);
+		conceptStateConversion.setProgramWorkflow(workflow);
+		conceptStateConversion.setProgramWorkflowState(state);
+		Context.getProgramWorkflowService().saveConceptStateConversion(conceptStateConversion);
+
+		handle(newDeleteRequest(getURI() + "/" + conceptStateConversion.getUuid(), new Parameter("purge", "true")));
+
+		assertNull(Context.getProgramWorkflowService().getConceptStateConversionByUuid(conceptStateConversion.getUuid()));
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java:122]: How do i tell that the above was not null before the delete request?

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+				"{\"concept\": \"" + concept.getUuid() + "\",\"programWorkflow\": \"" + workflow.getUuid()
+						+ "\",\"programWorkflowState\": \"" + state.getUuid() + "\"}";
+
+		SimpleObject newStateConversion = deserialize(handle(newPostRequest(getURI(), json)));
+
+		assertNotNull(newStateConversion);
+		String uuid = newStateConversion.get("uuid");
+
+		ConceptStateConversion createdStateConversion = Context.getProgramWorkflowService().getConceptStateConversionByUuid(uuid);
+		assertEquals(workflow.getUuid(), createdStateConversion.getProgramWorkflow().getUuid());
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java:95]: Can you also assert that the total count has increased?

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java (hunk 3 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/493/commits/3c9fee372101921ec32a3d3103b035e190ec4a33

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java (hunk 4 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java:101]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/493/commits/a2fa31aba58d4a72235cd93c5293e50b2c05e0f4

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		conceptStateConversion.setConcept(concept);
+		conceptStateConversion.setProgramWorkflow(workflow);
+		conceptStateConversion.setProgramWorkflowState(state);
+		conceptStateConversion.setUuid("e4f2fc12-0ca5-47d2-91bb-a129dca3149b");
+
+		return conceptStateConversion;
+	}
+
+	@Override
+	public String getDisplayProperty() {
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:46]: Are you sure that this should be null?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:46]: ConceptStateConversion doesn't have a display property right now. I could set it to, for example, the concept's name if you like.

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:46]: Did you see its toString method?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/ConceptStateConversionResource2_0Test.java:46]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/493/commits/6e38678dd1e465d3460208779243085b5bcaf081

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+		ConceptStateConversion createdStateConversion = service.getConceptStateConversionByUuid(uuid);
+		assertEquals(countBefore + 1, service.getAllConceptStateConversions().size());
+		assertEquals(workflow.getUuid(), createdStateConversion.getProgramWorkflow().getUuid());
+		assertEquals(state.getUuid(), createdStateConversion.getProgramWorkflowState().getUuid());
+		assertEquals(concept.getUuid(), createdStateConversion.getProgramWorkflow().getConcept().getUuid());
+	}
+
+	@Test
+	public void shouldPurgeStateConversion() throws Exception {
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java:107]: Shouldn't you also assert the size before and after?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ConceptStateConversionController2_0Test.java:107]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/493/commits/49f799d5673da3551053b60b88e79e4b0a1b08be

**Pull request merged**